### DISCUSSION TITLE
How make single query with `Select` and/or `Where` by many-to-many relations with 4 tables and nested queries?

### DISCUSSION THREAD

**Discussion by tomasevich** [prisma/prisma#20135] [Q&A] [answered]
Hey,

## Task

Make one query to get single **Right** by one unknow **Role** and known **Rule**

## Problem

I don't know what **Role** the Account has and it should be found in a single request by `email` (which I know)

## Target

I want to implement a **Role Model** and put configurations in a **Database**

## Case

Every time a user makes a request to the server, the system:
- Defines it `email` (cookie, form, etc...)
- By `email` determines its **Role**
- Sets a known **Rule** for each route (in my case the **ruleID** is `1`)
- Return **Right** with flag `true` or `false`

## Shema

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = env("DB_URL")
}

model Account {
  id                Int           @id @default(autoincrement())
  email             String        @unique
  hash              String        @default("UNDEFINED_HASH")

  role              AccountRole   @relation(fields: [roleId], references: [id])
  roleId            Int           @default(1)
}

model AccountRole {
  id                Int           @id @default(autoincrement())
  name              String        @unique

  accounts          Account[]
  rights            AccountRight[]
}

model AccountRule {
  id                Int           @id @default(autoincrement())
  name              String        @unique

  rights            AccountRight[]
}

model AccountRight {
  flag              Boolean       @default(false)

  role              AccountRole   @relation(fields: [roleId], references: [id])
  roleId            Int           @default(1)
  rule              AccountRule   @relation(fields: [ruleId], references: [id])
  ruleId            Int           @default(1)

  @@id([roleId, ruleId])
}
```

## Alternative solution

> I try to make 2 queries for get one result

- 1. Find **Role ID** by known **Email**

```js
const email = 'tomasevich@mail.local' // <- from req.body | cookie | etc...
const finedAccount = await prisma.account.findFirst({
  where: { email },
  select: { role: { select: { id: true } } }
})

console.log(finedAccount?.role?.id) // 3
```

- 2. Find **Right FLAG** by known **Role ID** and **Rule ID**

```js
const ruleId = 1 // <- from code | input data | etc...
const finedRight = await prisma.accountRight.findFirst({
  where: { ruleId, roleId: finedAccount?.role?.id },
  select: { flag: true }
})

console.log(finedRight?.flag) // true
```

## Not workin solution

In this case, I'm trying to get the `flag` from a known **ruleID** by determining the **Role** from its `email`

```js
const finedRight = await prisma.accountRight.findMany({
  where: {
    ruleId: 1
  },
  select: {
    flag: true,
    role: {
      select: {
        name: true,
        accounts: {
          where: { email: 'tomasevich@mail.local' },
          select: { email: true }
        }
      }
    }
  }
})

console.log(finedRight)
```

> **.findMany()** My target in finedRight[4], other elements don't needed

```json
[
  { "flag": false, "role": { "name": "Anonim", "accounts": [] } },
  { "flag": false, "role": { "name": "New user", "accounts": [] } },
  { "flag": true, "role": { "name": "Confirmed user", "accounts": [] } },
  { "flag": true, "role": { "name": "Admin", "accounts": [ { "email": "tomasevich@mail.local" } ] } }
]
```

> **.findFirst()** Return not that i want

```json
{ "flag": false, "role": { "name": "Anonim", "accounts": [] } }
```

## Where im looking for

- #18867 almost, but in this case only 2 nested query - i have 3 (i think)

**Comment by ludralph** [comment]
Hi @tomasevich 👋 

Can you please clarify a few things? If I understand your question correctly, you are trying to achieve the following:

1. Given an email address, determine the Role associated with that email.
2. Based on the determined Role and a known Rule ID, retrieve the flag value from the AccountRight model.
3. You want to accomplish this task in a single query. You have attempted to use two separate queries, one to find the Role ID based on the email and another to find the flag value based on the Role ID and Rule ID. However, you are looking for an alternative solution that combines these two queries into a single query.

Is my assumption correct?

**Reply by tomasevich** [reply]
Hey @ludralph, thank you for taking the time

Yeap x3

**Reply by tomasevich** [reply, marked as answer]
@ludralph i think i found solution 👇

```js
const finedRight = await prisma.accountRight.findFirst({
  where: {
    ruleId: 1, // Known Rule ID
    role: { // Unknown role
      is: { // Try to find it...
        accounts: { // ... in accounts
          some: { // one or more
            email: 'tomasevich@mail.local' // Known Email
          }
        }
      }
    }
  },
  select: {
    flag: true, // Select FLAG
    role: {
      select: {
        name: true, // Select founded Role NAME
        accounts: {
          select: { email: true } // Select founded EMAIL
        }
      }
    }
  }
})

console.log(finedRight)

// {"flag":true,"role":{"name":"Admin","accounts":[{"email":"tomasevich@mail.local"}]}}
```
